### PR TITLE
mprintf: reduce include overhead\nproper declare symbols with C linka…

### PR DESCRIPTION
**Pull request by slaff**
…ge\nfix crash when size is smaller than arg_size for %d %u\n fix padding for %d and %f\nfix buffer overflow for %d %f\n add vsnprintf equivalent \n add putch equivalent

Rebase of https://github.com/SmingHub/Sming/pull/748

### CODE REVIEW

File: Sming/system/stringconversion.cpp [hunk 1]
```diff
@@ -64,17 +70,22 @@ char* ultoa_wp(unsigned long val, char* buffer, unsigned int base, int width, ch
...
 }
 
+char *dtostrf(double floatVar, int minStringWidthIncDecimalPoint, int numDigitsAfterDecimal, char *outputBuffer)
+{
+	return dtostrf_p(floatVar, minStringWidthIncDecimalPoint, numDigitsAfterDecimal, outputBuffer, ' ');
+}
 // Author zitron: http://forum.arduino.cc/index.php?topic=37391#msg276209
 // modified by ADiea: remove dependencies strcat, floor, round; reorganize+speedup code
-char *dtostrf(double floatVar, int minStringWidthIncDecimalPoint, int numDigitsAfterDecimal, char *outputBuffer)
+extern char *dtostrf_p(double floatVar, int minStringWidthIncDecimalPoint, int numDigitsAfterDecimal, char *outputBuffer, char pad)
```
Comment (slaff) [inline Sming/system/stringconversion.cpp]: @ADiea why you have `extern` here?!

File: Sming/system/m_printf.cpp [hunk 2]
```diff
@@ -7,7 +7,8 @@ Descr: embedded very simple version of printf with float support
 */
 
 #include <stdarg.h>
-#include "../../SmingCore/SmingCore.h"
+//#include "../../SmingCore/SmingCore.h"
```
Comment (slaff) [inline Sming/system/m_printf.cpp]: @ADiea If you want to use only osapi.h then please remove the commented include above.

**Comment by anakod** [comment]
Is it ready?

**Comment by anakod** [comment]
I'm very like it and think it should be merged as soon as possible.

**Pull request merged**